selfdrive/car/subaru/carcontroller.py

```diff
@@ -46,6 +72,40 @@ def update(self, CC, CS, now_nanos):
 
       self.apply_steer_last = apply_steer
 
+    ### LONG ###
+
+    cruise_rpm = 0
+    cruise_throttle = 0
+
+    brake_cmd = False
+    brake_value = 0
+
+    if self.CP.openpilotLongitudinalControl:
+
+      gas, brake = compute_gb(actuators.accel)
+
+      if CC.longActive and brake > 0:
+        brake_value = clip(int(brake * CarControllerParams.BRAKE_SCALE), CarControllerParams.BRAKE_MIN, CarControllerParams.BRAKE_MAX)
+        brake_cmd = True
+
+      # AEB passthrough
+      if CC.enabled and CS.aeb:
+        brake_cmd = False
+
+      if CC.longActive and gas > 0:
+        # limit min and max values
+        cruise_throttle = clip(int(CarControllerParams.THROTTLE_BASE + (gas * CarControllerParams.THROTTLE_SCALE)), CarControllerParams.THROTTLE_MIN, CarControllerParams.THROTTLE_MAX)
+        cruise_rpm = clip(int(CarControllerParams.RPM_BASE + (gas * CarControllerParams.RPM_SCALE)), CarControllerParams.RPM_MIN, CarControllerParams.RPM_MAX)
+        # hysteresis
+        cruise_throttle, self.throttle_steady = accel_hysteresis(cruise_throttle, self.throttle_steady)
+        cruise_rpm, self.rpm_steady = accel_hysteresis(cruise_rpm, self.rpm_steady)
+
+        # slow down the signals change
+        cruise_throttle = clip(cruise_throttle, self.cruise_throttle_last - CarControllerParams.THROTTLE_DELTA, self.cruise_throttle_last + CarControllerParams.THROTTLE_DELTA)
+        cruise_rpm = clip(cruise_rpm, self.cruise_rpm_last - CarControllerParams.RPM_DELTA, self.cruise_rpm_last + CarControllerParams.RPM_DELTA)
```

Comment:
is rate limiting needed?

Reply:
the rate limits are set by observing stock signals and to smooth out rapid changes and reduce jerkiness like hysteresis. I would be interested if you have any suggestions on how to to model the openpilot long output signals to more closely follow the stock signals.